### PR TITLE
Add Python 3 support

### PR DESCRIPTION
**Pull request by ivuk**
As discussed in issue #2, here's the pull request that brings in Python 3 compatibility.

Note that I expect TravisCI run to fail on this, since `future` [supports Python 3.3 and later](http://python-future.org/faq.html#which-versions-of-python-does-python-future-support). If 3.2 is required, we could swap the `future` package for the `six` package, that one should be compatible with nearly every Python version.

### CODE REVIEW

**Comment by clusterfudge** [comment]
This does fail on 3.2, however our travisci config does not yet fail on unit test errors (which is where this error is presenting). I'm fine dropping 3.2 support, as that's seems to be a common enough choice when trying to make a 2.7 + >=3.3 compatible codebase (for codebases I respect, anyways :smile: , see https://github.com/ipython/ipython/pull/4002

I'm going to do a little more local testing, then i'll merge and fix the travis-ci stuff in post. 

Thanks for all the effort here!

**Comment by clusterfudge** [comment]
:disappointed: 

So, my (extremely thorough) tests have produced some less-than-stellar results. They are not conclusive, but more than sufficient to make me do a double take.

My tests consisted of timing the unit test suite for 1k iterations in serial. It's possible that there are deficiencies in the runner or the xml serialization library at play here, so it may be helpful to just create a stress test class to prove out py2 vs py3 performance.

TLDR; the python2 tests were almost identical for master vs this change, but the python3 tests took > 8x as long to run. As a python2 consumer of the lib, that doesn't affect me terribly, but it seems like a real gotcha for python3 devs.

**Comment by clusterfudge** [comment]
I have committed the change to master that will trigger build failures on unit test failures, and officially drops support for 3.2. Feel free to rebase.

**Comment by ivuk** [comment]
Rebased. :)

The performance issue sounds odd to me, though I haven't thoroughly checked out the code. A quick run with `cProfile` left me nowhere, I'll try to poke around some more later. If you have any ideas, I'm all ears.

**Comment by ivuk** [comment]
I've been trying to poke around with some flame graphs and PycallGraph, and I think it would help if we swap the `python-future` for `six` module quite a bit. Can you share the exact example you used for the 1k runs of the unit test suite? I've got nothing definitive at this time, but it looks a lot better to me.

I can also push this trivial changes to a new branch, say `py3-support-six`, so you could give it a go as well, if you want.

**Comment by ivuk** [comment]
I've pushed the changes to a [new branch](https://github.com/ivuk/adapt/tree/py3-support-six), give it a go when you find the time. :)

**Comment by clusterfudge** [comment]
so, upon further research, this doesn't seem to be an issue in adapt as much as it is in the test framework. These changes do not have a dramatic impact on performance, and I'm happy to merge. Do you have a preference which branch gets pulled in?

**Comment by ivuk** [comment]
Well, after trying both `six` and `future`, it seems to me that there are some advantages to using `six` - the module is smaller (`pip install future` pulls in ~1,6MB), PyCallGraph generates a smaller image for the same thing (likewise for the flame graph), and runtimes with cProfile seem to be a shade quicker in average. All of this are just "rule of thumb" measures, though. But since we are using just the `xrange` stuff, it might make sense to switch over to `six`. And with `six`, Python 3.2 support should work as well.

I'd love to hear more about your findings with regards to testing framework, if you can spare the time.

**Comment by clusterfudge** [comment]
Sounds good. You can open a new pr with the new branch, or you can rebase your changes off this branch. I'll leave that up to you.

As for my findings, I don't see a lot of value digging into it very deeply. My  initial "perf test suite" was running the tests 1k times. My new suite is running the single intent parser example 100 times. The disparity between py2 and py3 doesn't appear in the latter test, and I'm happy to chalk it up to the test framework.

Thanks for all your effort!

**Comment by ivuk** [comment]
Rebased, this should be it. With the switch to `six` module, Python 3.2 support could be reinstated, I think. :)

**Comment by clusterfudge** [comment]
Thanks for all the effort, this should go into the next release (since we're adding py3 support, i'll make that v0.2.0)

**Comment by ivuk** [comment]
No worries, keep up the good work! :)
